### PR TITLE
Task#11034 Detail Shopping List Screen

### PR DESCRIPTION
**Pull request by huybnd-1817**
### Issue Link :link:
https://edu-redmine.sun-asterisk.vn/issues/11034

### Goals :soccer:
https://marvelapp.com/4j8ce52/screen/55916912
Create UI for detail shopping list screen
- Recipe's name label
- Recipe's image
- Number of servings label
- Increase / Decrease number of servings button
- Back button
- List of ingredients using UITableView
- Remove button

Create a UI action
- Click increase / decrease number of servings 
- Click the ingredient which the user has bought
- Click remove button to delete this from shopping list
- Back to previous screen

Fetch data from local database and display to the UI
- Recipe ingredients' amount and number of servings

Edit and save data
- Update ingredients and number of servings

### Implementation Details :construction:
https://youtu.be/5ByjJFFPcT8

### CODE REVIEW

File: CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift [hunk 1]
```diff
@@ -40,4 +40,26 @@ extension IngredientMO {
...
         return ingredient
     }
+    
+    static func updateIngredient(ingredientName: String, isBought: Bool) {
+        let fetchRequest: NSFetchRequest<NSFetchRequestResult> = NSFetchRequest(entityName: "Ingredient")
+        let context = CoreDataManager.context
+        let predicate = NSPredicate(format: "ingredientName == %@", ingredientName)
+        fetchRequest.predicate = predicate
+        do {
+            let test = try context.fetch(fetchRequest)
```
Comment (longvt96) [inline CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift]: đổi tên biến `test` rõ ra nhé

File: CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift [hunk 2]
```diff
@@ -40,4 +40,26 @@ extension IngredientMO {
         print("Insert ingredient with name: \(ingredient.ingredientName ?? "") successful")
         return ingredient
     }
+    
+    static func updateIngredient(ingredientName: String, isBought: Bool) {
```
Comment (hungdt-1518) [inline CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift]: sử dụng hàm này thông qua singleton, ko nên để hàm static như thế này

File: CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift [hunk 3]
```diff
@@ -40,4 +40,26 @@ extension IngredientMO {
...
+        do {
+            let test = try context.fetch(fetchRequest)
+            if test.count == 1 {
+                let objectUpdate = test[0] as! NSManagedObject
+                objectUpdate.setValue(isBought, forKey: "isBought")
+                do {
+                    try context.save()
+                    print("Update ingredient successful")
+                } catch {
+                    print(error)
```
Comment (hungdt-1518) [inline CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift]: ko xử lý lỗi ?

Comment (tgdong2296) [reply CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift]: có force unwrap kìa

Comment (hungdt-1518) [reply CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift]: @tgdong2296 ấn crush chưa sao t ko thấy nhỉ

File: CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift [hunk 4]
```diff
@@ -40,4 +42,23 @@ extension IngredientMO {
         print("Insert ingredient with name: \(ingredient.ingredientName ?? "") successful")
         return ingredient
     }
+    
+    func updateIngredient(ingredientName: String, isBought: Bool) {
+        let fetchRequest: NSFetchRequest<NSFetchRequestResult> = NSFetchRequest(entityName: "Ingredient")
```
Comment (tienpx3010) [inline CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift]:
```suggestion
        let fetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: "Ingredient")
```

File: CuisineRecipe/CuisineRecipe/Scenes/DetailShoppingList/DetailShoppingListViewController.swift [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        ingredientsTableView.tableFooterView = UITableViewHeaderFooterView()
+    }
+    
+    @IBAction func handleGoToRecipeTapped(_ sender: Any) {
+        let vc = DetailRecipeViewController.instantiate()
+        vc.recipeId = recipe?.id
+        navigationController?.pushViewController(vc, animated: true)
+    }
+    
+    fileprivate func fetchData() {
```
Comment (tienpx3010) [inline CuisineRecipe/CuisineRecipe/Scenes/DetailShoppingList/DetailShoppingListViewController.swift]:
```suggestion
    private func fetchData() {
```

File: CuisineRecipe/CuisineRecipe/Scenes/DetailShoppingList/DetailShoppingListViewController.swift [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        recipeImageView.setupGradientBarBottom(width: view.frame.width, height: recipeImageView.frame.height)
+        numOfServingsLabel.text = String(numsOfServings) + " servings"
+        sortedIngredients = ingredients.sorted {
+            guard let ingredientName1 = $0.ingredientName,
+            let ingredientName2 = $1.ingredientName else { return false }
+            return ingredientName1 < ingredientName2
+        }
+    }
+    
+    @IBAction func handleRemoveRecipeTapped(_ sender: Any) {
```
Comment (tienpx3010) [inline CuisineRecipe/CuisineRecipe/Scenes/DetailShoppingList/DetailShoppingListViewController.swift]: private

File: CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift [hunk 7]
```diff
@@ -40,4 +42,23 @@ extension IngredientMO {
         print("Insert ingredient with name: \(ingredient.ingredientName ?? "") successful")
         return ingredient
     }
+    
+    func updateIngredient(ingredientName: String, isBought: Bool) {
```
Comment (tienpx3010) [inline CuisineRecipe/CuisineRecipe/CoreData/IngredientMO+CoreDataProperties.swift]: -> Bool để kiểm tra success hay error

File: CuisineRecipe/CuisineRecipe/Scenes/DetailRecipe/DetailRecipeViewController.swift [hunk 8]
```diff
@@ -106,24 +106,24 @@ final class DetailRecipeViewController: UIViewController {
             !img.isEmpty,
             let imgUrl = img[0].hostedLargeUrl else { return }
         
-        guard !RecipeMO.recipeExists(recipeId: recipeId) else {
+        guard !RecipeMO.shared.recipeExists(recipeId: recipeId) else {
             showToast(message: "You have added this recipe to shopping list")
```
Comment (tienpx3010) [inline CuisineRecipe/CuisineRecipe/Scenes/DetailRecipe/DetailRecipeViewController.swift]: Bỏ vào Constants

**Pull request merged**